### PR TITLE
In-Memory: implement hybrid snapshot iter

### LINKED ISSUE #16684: Implement HybridEngineSnapshot::iterator_opt
## Implement HybridEngineSnapshot::iterator_opt 

Found this issue when running the tests, porting this fix from my TPCC e2e test branch.

### PR DESCRIPTION
**Pull request by afeinberg**
### What is changed and how it works?

Issue Number: Ref #16141 Close #16684

What's Changed:

```commit-message
Implements an iterator for HybridEngineSnapshot and fixes a bug in HybridEngineSnapshot::get_value_opt (memory-engine only supports the data cfs.)
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)
Tested during TPCC runs in a separate branch.
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note
None
-->

```release-note
None

```

### CODE REVIEW

File: components/hybrid_engine/src/snapshot.rs [hunk 1]
```diff
@@ -68,18 +66,36 @@ where
     EK: KvEngine,
     EC: RangeCacheEngine,
 {
-    type Iterator = HybridEngineIterator<EK, EC>;
+    // TODO (afeinberg): use `Either` instead of `Box<dyn Iterator>`
```
Comment (SpadeA-Tang) [inline components/hybrid_engine/src/snapshot.rs:69]: Why not implement Either directly so that we dont need Box.

Comment (afeinberg) [reply components/hybrid_engine/src/snapshot.rs]: done.

**Comment by afeinberg** [comment]
Addressed requests.

**Review by SpadeA-Tang — APPROVED** [review]

**Review by v01dstar — APPROVED** [review]
LGTM

**Review by tonyxuqqi — APPROVED** [review]

**Pull request merged**